src/components/Input/Input.jsx

```diff
@@ -0,0 +1,141 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import classNames from 'classnames';
+import { v4 as uuid } from 'uuid';
+import './Input.scss';
+
+const Input = (props) => {
+  const {
+    className,
+    id,
+    isDisabled,
+    label,
+    onBlur,
+    onChange,
+    onFocus,
+    placeholder,
+    type,
+    value,
+  } = props;
+
+  const handleChange = (event) => {
+    onChange(event);
+  }
+
+  const handleFocus = (event) => {
+    if (onFocus) onFocus(event);
+  }
+
+  const handleBlur = (event) => {
+    if (onBlur) onBlur(event);
+  }
+
+  const classes = classNames(
+    className,
+    'input',
+    // { @TODO Placeholder until size/color props are added.
+    //   [`font-size-${size}`]: size,
+    //   [`font-color-${color}`]: color,
+    // }
+  );
+
+  const inputId = id || uuid();
+
+  const renderInput = () => (
+    <input
+      className={classes}
+      disabled={isDisabled}
+      id={inputId}
+      onBlur={handleBlur}
+      onChange={handleChange}
+      onFocus={handleFocus}
+      placeholder={placeholder}
+      type={type}
+      value={value}
+    />
+  );
+
+  const renderLabel = () => {
+    if (!label) return;
+
+    return (
+      <label className="label" htmlFor={inputId}>
+        {label}<span className="font-color-danger font-size-sm">&nbsp;*</span>
+      </label>
+    );
+  }
+
+  return (
+    label ? (
+      <>
+        {renderLabel()}
+        {renderInput()}
+      </>
+    ) : (
+      <>
+        {renderInput()}
+      </>
+    )
+  );
+};
+
+Input.propTypes = {
+  /**
+   * Custom class to be added to standard input classes.
+   */
+  className: PropTypes.string,
+  /**
+   * The input's id attribute. We will use this to programmatically tie the input with its label.
+   */
+  id: PropTypes.string,
+  /**
+   * The input's disabled attribute
+   */
+  isDisabled: PropTypes.bool,
+  /**
+   * Determines if input is required or not. (Label will have an asterisk if required)
+   */
+  isRequired: PropTypes.bool,
+  /**
+   * Value for HTML <label> tag
+   */
+  label: PropTypes.string,
+  /**
+   * Callback function to call on blur event.
+   */
+  onBlur: PropTypes.func,
+  /**
+   * Callback function to call on change event.
+   */
+  onChange: PropTypes.func.isRequired,
+  /**
+   * Callback function to call on focus event..
+   */
+  onFocus: PropTypes.func,
+  /**
+   * The input placeholder attribute
+   */
+  placeholder: PropTypes.string,
+  /**
+   * The input 'type' value. Defaults to type 'text'.
+   */
+  type: PropTypes.oneOf(['text', 'password', 'email', 'tel', 'url', 'search']),
+  /**
+   * The text value of the input. Required since our Input is a controlled component.
+   */
+  value: PropTypes.string.isRequired,
+};
+
+Input.defaultProps = {
+  className: '',
+  id: undefined,
+  isDisabled: false,
+  isRequired: false,
+  label: undefined,
+  onBlur: undefined,
+  onFocus: undefined,
+  placeholder: '',
+  type: 'text',
+};
+
+export default Input;
```

Comment:
i see you're declaring the 2 sub-render functions inside the functional component itself. Is there an advantage to this versus declaring those functions outside of the main component and passing in the relevant info? In functional components, the function gets run every time the component rerenders. We probably don't want to redeclare these functions right?

Reply:
Interesting insight; hadn't realized there was a performance impact for doing this. 

I've always been a fan of nesting inside the parent component since they belong to only this component. So for me, it's more about clarity for the dev, as apposed to worrying about a small performance hit. Lastly, the react js docs seem to be for nesting as can be seen in this hooks example: https://reactjs.org/docs/hooks-overview.html

Either way we go on this, this is a pattern that we should establish now, and stick to it.

Reply:
@JCarmine so hooks definitely need to called within the parent function component. It's more the subcomponents (renderLabel and renderInput) in this case that can go outside the main parent component. To me, 1 file represents 1 component, so its ok if a file has 1 main component and several sub components defined at the top level (as long as the main component is always at the top or bottom of the file) (i prefer top)

Reply:
in addition, hooks that don't need to be in the parent component and only have to do with one of the sub components can go directly into the subcomponent

Reply:
I am with @ericyliu on this one. Typically I like to keep (sub)render functions inside the parent component scope, purely due to the ease of dev experience given that we can keep a single prop scope across the entire thing. In general I would opt for that pattern moving forward but this case is different because the specific render functions actually can be abstracted to their own components. The truth is we will need to re-use the form label for every single input type so it is worth doing that now and having it be agnostic of it's input altogether. 

I will work on updating this tonight and we can review in the morning.